// git-log-as-object/lib/Diff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const process_1 = require("./process");
const args = (commit) => [
    'diff-tree',
    '--no-commit-id',
    '--name-status',
    '-r',
    commit.fullHash,
];
function parse(rawData) {
    const files = rawData.split('\n').slice(0, -1).map(s => s.split('\t'));
    const diff = {
        added: new Set(),
        deleted: new Set(),
        modified: new Set(),
        touched: new Set(),
    };
    for (const [mode, path] of files) {
        diff.touched.add(path);
        switch (mode) {
            case 'A':
                diff.added.add(path);
                continue;
            case 'D':
                diff.deleted.add(path);
                continue;
            case 'M':
                diff.modified.add(path);
                continue;
        }
    }
    return diff;
}
var Diff;
(function (Diff) {
    Diff.make = (commit, dir) => process_1.wrapGitProcess(args(commit), dir).then(parse);
    Diff.makeSync = (commit, dir) => parse(process_1.wrapGitProcessSync(args(commit), dir));
})(Diff = exports.Diff || (exports.Diff = {}));
